clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -409,6 +409,24 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
                 instr_free(GLOBAL_DCONTEXT, &instr);
             }
         }
+        if (shard->prev_instr_.instr.addr != 0 /*first*/) {
+            instr_t prev_instr, curr_instr;
+            instr_init(GLOBAL_DCONTEXT, &curr_instr);
+            instr_init(GLOBAL_DCONTEXT, &prev_instr);
+            decode_from_copy(
+                GLOBAL_DCONTEXT, const_cast<app_pc>(shard->prev_instr_.instr.encoding),
+                reinterpret_cast<app_pc>(shard->prev_instr_.instr.addr), &prev_instr);
+            decode_from_copy(GLOBAL_DCONTEXT, const_cast<app_pc>(memref.instr.encoding),
```

Comment:
We're already decoding the cur instr for the "Branch target is not decodeable" check. We should decode only once and use it in both checks.

Reply:
The current instr is being decoded inside an else clause in "Branch target is not decodeable" error so it's possible we might not always see it by the time we want to use it for the Repeated syscall check.

Reply:
I mean: can we move both `decode_from_copy`s to a common location so that we need to do it only once per iteration?

Reply:
Ack on the approach but note that in "Branch target is not decodeable", it's the previous instr that's being decoded.

Reply:
How about always decoding the current instr in each iteration, and then assigning it to `shard->decoded_prev_instr` pointer so that the next iteration can use it?

I think it'll be good for overhead to extract the instr decoding to a common location, since we may have more and more checks that require decoding instrs.